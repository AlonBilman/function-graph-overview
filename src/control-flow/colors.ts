const defaultColorList = [
  // Node colors
  { name: "node.default", hex: "#d3d3d3" },
  { name: "node.entry", hex: "#48AB30" },
  { name: "node.exit", hex: "#AB3030" },
  { name: "node.throw", hex: "#ffdddd" },
  { name: "node.yield", hex: "#00bfff" },
  { name: "node.border", hex: "#000000" },
  { name: "node.highlight", hex: "#000000" },

  // Edge Colors
  { name: "edge.regular", hex: "#0000ff" },
  { name: "edge.consequence", hex: "#008000" },
  { name: "edge.alternative", hex: "#ff0000" },

  // Cluster Colors
  { name: "cluster.border", hex: "#ffffff" },
  { name: "cluster.with", hex: "#ffddff" },
  { name: "cluster.tryComplex", hex: "#ddddff" },
  { name: "cluster.try", hex: "#ddffdd" },
  { name: "cluster.finally", hex: "#ffffdd" },
  { name: "cluster.except", hex: "#ffdddd" },

  // Graph Colors
  { name: "graph.background", hex: "#ffffff" },
] as const;

export type ColorList = typeof defaultColorList;
export type ColorScheme = Record<ColorList[number]["name"], string>;
export function listToScheme(colors: ColorList): ColorScheme {
  const scheme = {} as ColorScheme;
  for (const { name, hex } of colors) {
    scheme[name] = hex;
  }
  return scheme;
}

const defaultColorScheme = listToScheme(defaultColorList);

export function getDefaultColorList(): ColorList {
  return structuredClone(defaultColorList);
}

export function getDefaultColorScheme(): ColorScheme {
  return structuredClone(defaultColorScheme);
}

export function serializeColorList(colorList: ColorList): string {
  return JSON.stringify({ version: 1, scheme: colorList });
}

export function deserializeColorList(data: string): ColorList {
  const { version, scheme } = JSON.parse(data);
  if (version !== 1) {
    throw new Error(`Invalid scheme version: ${version}`);
  }
  for (const { hex } of scheme) {
    if (!hex.match(/^#[0-9a-fA-F]+$/)) {
      throw new Error(`Invalid color: ${hex}`);
    }
  }
  return scheme;
}
