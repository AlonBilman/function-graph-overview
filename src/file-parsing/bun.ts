import type Parser from "web-tree-sitter";
import { type Language, supportedLanguages } from "../control-flow/cfg.ts";
import { initializeParser } from "../parser-loader/bun.ts";
import { makeIterFunctions } from "./file-parsing.ts";
export { getLanguage, fileTypes } from "./file-parsing.ts";

const parsers: { [language in Language]: Parser } = Object.fromEntries(
  await (async () => {
    const parsers = [];
    for (const language of supportedLanguages) {
      parsers.push([language, (await initializeParser(language)).parser]);
    }
    return parsers;
  })(),
);

export const iterFunctions = makeIterFunctions(parsers);
