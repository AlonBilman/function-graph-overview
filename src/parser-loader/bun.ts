import { Parser, Language as TreeSitterLanguage } from "web-tree-sitter";
import { type Language, languageDefinitions } from "../control-flow/cfg.ts";

export async function initializeParser(
  language: Language,
): Promise<{ parser: Parser; language: TreeSitterLanguage }> {
  await Parser.init();

  const parserLanguage = await TreeSitterLanguage.load(
    languageDefinitions[language].wasmPath,
  );
  const parser = new Parser();
  parser.setLanguage(parserLanguage);
  return { parser, language: parserLanguage };
}
