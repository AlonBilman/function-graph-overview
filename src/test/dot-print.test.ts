import { describe, expect, test } from "bun:test";

import { getLightColorScheme } from "../control-flow/colors.ts";
import { applyTheme } from "../dot-cfg/dot-print.ts";

describe("Theme DOT-CFG Samples", async () => {
  test("Theme CFG", () => {
    const result = applyTheme(
      `
      digraph {
          entry [class="entry"]
          return1 [class="exit"]
          return2 [class="exit"]
          return4_2 [class="exit"]
          stmt2 [height=2]
          stmt3 [height=2]
          entry -> stmt1
          stmt1 -> return1 [class="consequence"]
          stmt1 -> stmt2 [class="alternative"]
          stmt2 -> return2 [class="consequence"]
          stmt2 -> stmt3 [class="alternative"]
          stmt3 -> stmt4_1
          stmt4_1 -> stmt4_2
          stmt4_2 -> return4_2 [class="consequence"]
          stmt4_2 -> stmt4_4 [class="alternative"]
          stmt4_4 -> stmt4_1 [dir="back"]
          stmt4_4 [height=3]
      }
      `,
      getLightColorScheme(),
    );
    expect(result).toMatchSnapshot();
  });

  test("Theme CFG with clusters", () => {
    const result = applyTheme(
      `
      digraph {
          subgraph cluster_tryComplex {
              class=tryComplex
              subgraph cluster_try {
                  class=try
                  TRY [label=try]
              }
              subgraph cluster_except {
                  class=except
                  EXCEPT [label="except/catch"]
              }
              subgraph cluster_else {
                  ELSE [label=else]
              }
              subgraph cluster_finally {
                  class=finally
                  FINALLY [label=finally]
              }
          }
          ENTRY -> TRY
          TRY -> EXCEPT [class=exception]
          TRY -> ELSE
          ELSE -> FINALLY
          EXCEPT -> FINALLY
          FINALLY -> EXIT
      }
    `,
      getLightColorScheme(),
    );
    expect(result).toMatchSnapshot();
  });
});
