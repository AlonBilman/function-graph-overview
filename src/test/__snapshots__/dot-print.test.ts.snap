// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Theme DOT-CFG Samples > Theme CFG 1`] = `
"digraph {
  bgcolor = "#ffffff";
  edge [
    penwidth = 1;
    color = "#0000ff";
    headport = "n";
    tailport = "s";
  ];
  node [
    label = "";
    style = "filled";
    shape = "box";
    class = "default";
    fillcolor = "#d3d3d3";
    color = "#000000";
    fontname = "sans-serif";
    height = 0.3;
  ];
  "entry" [
    class = "entry";
    shape = "invhouse";
    fillcolor = "#48AB30";
    color = "#000000";
    height = 0.5;
  ];
  "return1" [
    class = "exit";
    shape = "house";
    fillcolor = "#AB3030";
    color = "#000000";
    height = 0.5;
  ];
  "return2" [
    class = "exit";
    shape = "house";
    fillcolor = "#AB3030";
    color = "#000000";
    height = 0.5;
  ];
  "return4_2" [
    class = "exit";
    shape = "house";
    fillcolor = "#AB3030";
    color = "#000000";
    height = 0.5;
  ];
  "stmt2" [
    height = 0.6;
    label = "";
    style = "filled";
    shape = "box";
    class = "default";
    fillcolor = "#d3d3d3";
    color = "#000000";
  ];
  "stmt3" [
    height = 0.6;
    label = "";
    style = "filled";
    shape = "box";
    class = "default";
    fillcolor = "#d3d3d3";
    color = "#000000";
  ];
  "stmt4_4" [
    height = 0.8999999999999999;
    label = "";
    style = "filled";
    shape = "box";
    class = "default";
    fillcolor = "#d3d3d3";
    color = "#000000";
  ];
  "entry" -> "stmt1" [
    class = "regular";
    color = "#0000ff";
  ];
  "stmt1" -> "return1" [
    class = "consequence";
    color = "#008000";
  ];
  "stmt1" -> "stmt2" [
    class = "alternative";
    color = "#ff0000";
  ];
  "stmt2" -> "return2" [
    class = "consequence";
    color = "#008000";
  ];
  "stmt2" -> "stmt3" [
    class = "alternative";
    color = "#ff0000";
  ];
  "stmt3" -> "stmt4_1" [
    class = "regular";
    color = "#0000ff";
  ];
  "stmt4_1" -> "stmt4_2" [
    class = "regular";
    color = "#0000ff";
  ];
  "stmt4_2" -> "return4_2" [
    class = "consequence";
    color = "#008000";
  ];
  "stmt4_2" -> "stmt4_4" [
    class = "alternative";
    color = "#ff0000";
  ];
  "stmt4_1" -> "stmt4_4" [
    dir = "back";
    class = "regular";
    color = "#0000ff";
    penwidth = 2;
    headport = "s";
    tailport = "n";
  ];
}"
`;

exports[`Theme DOT-CFG Samples > Theme CFG with clusters 1`] = `
"digraph {
  bgcolor = "#ffffff";
  edge [
    penwidth = 1;
    color = "#0000ff";
    headport = "n";
    tailport = "s";
  ];
  node [
    label = "";
    style = "filled";
    shape = "box";
    class = "default";
    fillcolor = "#d3d3d3";
    color = "#000000";
    fontname = "sans-serif";
    height = 0.3;
  ];
  subgraph "cluster_tryComplex" {
    class = "tryComplex";
    graph [
      penwidth = 0;
      class = "tryComplex";
      color = "#ffffff";
      bgcolor = "#ddddff";
    ];
    subgraph "cluster_try" {
      class = "try";
      graph [
        penwidth = 0;
        class = "try";
        color = "#ffffff";
        bgcolor = "#ddffdd";
      ];
      "TRY" [
        label = "try";
        style = "filled";
        shape = "box";
        class = "default";
        fillcolor = "#d3d3d3";
        color = "#000000";
        height = 0.3;
      ];
    }
    subgraph "cluster_except" {
      class = "except";
      graph [
        penwidth = 0;
        class = "except";
        color = "#ffffff";
        bgcolor = "#ffdddd";
      ];
      "EXCEPT" [
        label = "except/catch";
        style = "filled";
        shape = "box";
        class = "default";
        fillcolor = "#d3d3d3";
        color = "#000000";
        height = 0.3;
      ];
    }
    subgraph "cluster_else" {
      "ELSE" [
        label = "else";
        style = "filled";
        shape = "box";
        class = "default";
        fillcolor = "#d3d3d3";
        color = "#000000";
        height = 0.3;
      ];
    }
    subgraph "cluster_finally" {
      class = "finally";
      graph [
        penwidth = 0;
        class = "finally";
        color = "#ffffff";
        bgcolor = "#ffffdd";
      ];
      "FINALLY" [
        label = "finally";
        style = "filled";
        shape = "box";
        class = "default";
        fillcolor = "#d3d3d3";
        color = "#000000";
        height = 0.3;
      ];
    }
  }
  "ENTRY" -> "TRY" [
    class = "regular";
    color = "#0000ff";
  ];
  "TRY" -> "EXCEPT" [
    class = "exception";
    style = "invis";
    headport = "e";
    tailport = "w";
  ];
  "TRY" -> "ELSE" [
    class = "regular";
    color = "#0000ff";
  ];
  "ELSE" -> "FINALLY" [
    class = "regular";
    color = "#0000ff";
  ];
  "EXCEPT" -> "FINALLY" [
    class = "regular";
    color = "#0000ff";
  ];
  "FINALLY" -> "EXIT" [
    class = "regular";
    color = "#0000ff";
  ];
}"
`;
